lib/ansible/parsing/vault/editor.py

```diff
@@ -0,0 +1,413 @@
+                    chunk_len = random.randint(max_chunk_len // 2, max_chunk_len)
+                    data = os.urandom(chunk_len)
+
+                    for _ in range(0, file_len // chunk_len):
+                        fh.write(data)
+                    fh.write(data[:file_len % chunk_len])
+
+                    # FIXME remove this assert once we have unittests to check its accuracy
+                    if fh.tell() != file_len:
+                        raise AnsibleAssertionError()
+
+                    os.fsync(fh)
+
+    def _shred_file(self, tmp_path):
+        """
+        Securely destroy a decrypted file
+
+        Note standard limitations of GNU shred apply (For flash, overwriting would have no effect
+        due to wear leveling; for other storage systems, the async kernel->filesystem->disk calls never
+        guarantee data hits the disk; etc). Furthermore, if your tmp dirs is on tmpfs (ramdisks),
+        it is a non-issue.
+
+        Nevertheless, some form of overwriting the data (instead of just removing the fs index entry) is
+        a good idea. If shred is not available (e.g. on windows, or no core-utils installed), fall back on
+        a custom shredding method.
+        """
+
+        if not os.path.isfile(tmp_path):
+            # file is already gone
+            return
+
+        try:
+            r = subprocess.call(['shred', tmp_path])
+        except (OSError, ValueError):
+            # shred is not available on this system, or some other error occurred.
+            # ValueError caught because macOS El Capitan is raising an
+            # exception big enough to hit a limit in python2-2.7.11 and below.
+            # Symptom is ValueError: insecure pickle when shred is not
+            # installed there.
+            r = 1
+
+        if r != 0:
+            # we could not successfully execute unix shred; therefore, do custom shred.
+            self._shred_file_custom(tmp_path)
+
+        os.remove(tmp_path)
+
+    def _edit_file_helper(self, filename, secret, existing_data=None, force_save=False, vault_id=None):
+
+        # Create a tempfile
+        root, ext = os.path.splitext(os.path.realpath(filename))
+        fd, tmp_path = tempfile.mkstemp(suffix=ext, dir=C.DEFAULT_LOCAL_TMP)
+
+        cmd = self._editor_shell_command(tmp_path)
+        try:
+            if existing_data:
+                self.write_data(existing_data, fd, shred=False)
+        except Exception:
+            # if an error happens, destroy the decrypted file
+            self._shred_file(tmp_path)
+            raise
+        finally:
+            os.close(fd)
+
+        try:
+            # drop the user into an editor on the tmp file
+            subprocess.call(cmd)
+        except Exception as e:
+            # if an error happens, destroy the decrypted file
+            self._shred_file(tmp_path)
+            raise AnsibleError(f'Unable to execute the command "{" ".join(cmd)}": {to_native(e)}')
```

Comment:
`to_native` is unnecessary now with the f string. You could raise the AnsibleError from the original exception:
```suggestion
            raise AnsibleError(f'Unable to execute the command "{" ".join(cmd)}": {e}') from e
```

Reply:
fixed in [4a27827](https://github.com/ansible/ansible/pull/80675/commits/4a27827f66f47692ecbff9058aedf38cd6d57d8a)